### PR TITLE
udev: introduce AlternativeNamesPolicy= setting

### PR DESCRIPTION
**Pull request by yuwata**
@jpirko Feel free to use, pick, or modify this.

### CODE REVIEW

**Comment by jpirko** [comment]
FWIW, this looks good to me. Tested, works fine. Thanks! Next time I'll wait one more week before I start to implement things in udev :)

File: man/systemd.link.xml [hunk 1]
```diff
@@ -356,6 +356,16 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>AlternativeNamesPolicy=</varname></term>
+        <listitem>
+          <para>An ordered, space-separated list of policies by which the interface's alternative names
```
Comment (poettering) [inline man/systemd.link.xml]: "ordered"? The order shouldn't really matter much, no? the policies listed here are independent of order, any order is fine, and effectively equivalent, no?

File: src/udev/net/link-config.c [hunk 2]
```diff
@@ -459,9 +461,50 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
...
+                                r = strv_extend(&altnames, n);
+                                if (r < 0)
+                                        return log_oom();
+                        }
+                }
+
+        if (new_name)
+                strv_remove(altnames, new_name);
+        strv_remove(altnames, old_name);
+        strv_uniq(altnames);
```
Comment (poettering) [inline src/udev/net/link-config.c:505]: maybe even do strv_sort() here, to make the order stable? (will the kernel maintain the order btw?)

**Comment by poettering** [comment]
hmm, one more thing. udev's net_id thing currently enforces the size limit and only sets tha various ID_NET_NAME_xyz fields only for names fitting into IFNAMSIZ. This should be updated to set the fields even if they violate IFNAMSIZ as long as they don't violate ALTIFNAMSIZ. But then we have to be more careful when reading these fields and check if the specified names are suitable as primary name, and ignore them if not.

(I figure one could claim a compat borkage if we start to export ID_NET_NAME_xyz fields that are longer than IFNAMSIZ, but I think that'd be minor and would risk it.)

File: man/systemd.link.xml (hunk 1 above)
Comment (yuwata) [reply man/systemd.link.xml]: This is a copy and paste mistake. Dropped.

File: src/udev/net/link-config.c (hunk 2 above)
Comment (yuwata) [reply src/udev/net/link-config.c:505]: Why the order is important?

**Comment by yuwata** [comment]
@poettering Thank you for the comment. I've force-pushed a revised version. PTAL.

> Hmm, do we have a somewhat reasonable fall back for older kernels? Most likely we should simply not apply any alternative names if the concept does not exist on the running kernel. (needs to be documented I figure)

In the revised version, now the error `-EOPNOTSUPP` in setting alternative names is ignored. Also, document is updated.

> hmm, one more thing. udev's net_id thing currently enforces the size limit and only sets tha various ID_NET_NAME_xyz fields only for names fitting into IFNAMSIZ. This should be updated to set the fields even if they violate IFNAMSIZ as long as they don't violate ALTIFNAMSIZ. But then we have to be more careful when reading these fields and check if the specified names are suitable as primary name, and ignore them if not.
> 
> (I figure one could claim a compat borkage if we start to export ID_NET_NAME_xyz fields that are longer than IFNAMSIZ, but I think that'd be minor and would risk it.)

Now, the length of ID_NET_NAME_xxx is extended to ALTIFNAMSIZ.

**Comment by jpirko** [comment]
> I like very much. Just some minor notes on ordering...
> 
> Hmm, do we have a somewhat reasonable fall back for older kernels? Most likely we should simply not apply any alternative names if the concept does not exist on the running kernel. (needs to be documented I figure)

The only way you know kernel does not support this is that you add altname and you receive -eopnotsupp. I would just check if -eopnotsupp was returned from netlink and in that case do nothing.

File: src/udev/net/link-config.c (hunk 2 above)
Comment (jpirko) [reply src/udev/net/link-config.c:505]: > maybe even do strv_sort() here, to make the order stable? (will the kernel maintain the order btw?)

Yes, kernel is going to maintain the order.

Comment (poettering) [reply src/udev/net/link-config.c:505]: order is not important. Just easier to look at if people type "ip link" I'd figure?

**Comment by poettering** [comment]
LGTM. sorting the alternative names we apply would make sense to me still, but even without looks excellent

**Pull request merged**